File: mqtt_cpp/mqtt_sender.cpp
#include <mosquitto.h>
#include <iostream>
#include <string>
#include <chrono>
#include <thread>
#include <signal.h>
#include <json/json.h>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <queue>
#include <mutex>
#include <condition_variable>

// 全局变量
volatile bool running_ = true;
struct mosquitto* mosq = nullptr;

// 消息队列
std::queue<std::string> mqtt_message_queue;
std::mutex queue_mutex;
std::condition_variable queue_cv;

// 信号处理函数
void signal_handler(int signum) {
    std::cout << "\n接收到信号 " << signum << "，正在停止程序..." << std::endl;
    running_ = false;
    queue_cv.notify_all();
}

// MQTT连接回调
void mqtt_connect_callback(struct mosquitto* mosq, void* userdata, int result) {
    // 静默模式，不打印连接信息
}

// MQTT发布回调
void mqtt_publish_callback(struct mosquitto* mosq, void* userdata, int mid) {
    // 静默模式，不打印发布信息
}

// MQTT断开连接回调
void mqtt_disconnect_callback(struct mosquitto* mosq, void* userdata, int result) {
    // 静默模式，不打印断开信息
}

// 获取当前时间戳（毫秒）
int64_t get_current_timestamp_ms() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now().time_since_epoch()).count();
}

// 获取格式化的时间戳字符串
std::string get_formatted_timestamp() {
    auto now = std::chrono::system_clock::now();
    auto time_t = std::chrono::system_clock::to_time_t(now);
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()) % 1000;
    
    std::stringstream ss;
    ss << std::put_time(std::localtime(&time_t), "%H:%M:%S");
    ss << '.' << std::setfill('0') << std::setw(3) << ms.count();
    return ss.str();
}

// 发送MQTT消息
bool send_mqtt_message(const std::string& message, const std::string& topic = "cloud/vehicle/control") {
    if (!mosq) {
        return false;
    }
    
    int result = mosquitto_publish(mosq, nullptr, topic.c_str(), message.length(), message.c_str(), 1, false);
    
    return (result == MOSQ_ERR_SUCCESS);
}

// MQTT发送线程函数
void mqtt_sender_thread() {
    std::cout << "MQTT发送线程已启动" << std::endl;
    
    while (running_) {
        std::string message;
        {
            std::unique_lock<std::mutex> lock(queue_mutex);
            queue_cv.wait(lock, []{ return !mqtt_message_queue.empty() || !running_; });
            
            if (!running_) break;
            
            if (!mqtt_message_queue.empty()) {
                message = mqtt_message_queue.front();
                mqtt_message_queue.pop();
            }
        }
        
        if (!message.empty()) {
            send_mqtt_message(message);
        }
    }
    
    std::cout << "MQTT发送线程已停止" << std::endl;
}

// CSV写入线程函数
void csv_writer_thread(std::ofstream& csv_file, std::queue<std::vector<std::string>>& csv_queue, std::mutex& csv_mutex, std::condition_variable& csv_cv) {
    std::vector<std::string> data;
    
    while (running_) {
        {
            std::unique_lock<std::mutex> lock(csv_mutex);
            csv_cv.wait(lock, [&]{ return !csv_queue.empty() || !running_; });
            
            if (!running_) break;
            
            if (!csv_queue.empty()) {
                data = csv_queue.front();
                csv_queue.pop();
            }
        }
        
        if (!data.empty()) {
            // 写入CSV
            for (size_t i = 0; i < data.size(); ++i) {
                if (i > 0) csv_file << ",";
                csv_file << data[i];
            }
            csv_file << "\n";
            csv_file.flush();
        }
    }
}

class MQTTSender {
private:
    std::string broker_;
    int port_;
    std::string topic_;
    int sequence_id_;
    double interval_;
    std::ofstream csv_file_;
    std::queue<std::vector<std::string>> csv_queue_;
    std::mutex csv_mutex_;
    std::condition_variable csv_cv_;
    std::thread csv_thread_;
    
public:
    MQTTSender(const std::string& broker = "hellorobotaxi.cn", int port = 11883, const std::string& topic = "cloud/vehicle/control") 
        : broker_(broker), port_(port), topic_(topic), sequence_id_(0), interval_(0.01) {
        
        // 创建日志目录
        system("mkdir -p logs");
        
        // 创建带时间戳的日志目录
        auto now = std::chrono::system_clock::now();
        auto time_t = std::chrono::system_clock::to_time_t(now);
        std::stringstream ss;
        ss << "logs/sender_" << std::put_time(std::localtime(&time_t), "%Y%m%d_%H%M%S");
        std::string log_dir = ss.str();
        
        // 确保目录唯一
        int counter = 0;
        std::string final_log_dir = log_dir;
        while (system(("test -d " + final_log_dir).c_str()) == 0) {
            counter++;
            final_log_dir = log_dir + "_" + std::to_string(counter);
        }
        
        system(("mkdir -p " + final_log_dir).c_str());
        
        // 创建CSV文件
        csv_file_.open(final_log_dir + "/sender_data.csv");
        csv_file_ << "sequence_id,a1_timestamp\n";
        
        std::cout << "日志目录: " << final_log_dir << std::endl;
        
        // 启动CSV写入线程
        csv_thread_ = std::thread(csv_writer_thread, std::ref(csv_file_), std::ref(csv_queue_), std::ref(csv_mutex_), std::ref(csv_cv_));
    }
    
    ~MQTTSender() {
        running_ = false;
        csv_cv_.notify_all();
        if (csv_thread_.joinable()) {
            csv_thread_.join();
        }
        if (csv_file_.is_open()) {
            csv_file_.close();
        }
    }
    
    void send_message() {
        if (!mosq) {
            return;
        }
        
        try {
            // 获取时间戳并发送消息
            sequence_id_++;
            int64_t a1_timestamp = get_current_timestamp_ms();
            
            // 创建JSON消息
            Json::Value message;
            message["sequence_id"] = sequence_id_;
            message["steering_angle_enable"] = true;
            message["steering_angle"] = 0.0;
            message["target_acceleration_enable"] = true;
            message["target_acceleration"] = 0.0;
            message["indicator_left_enable"] = false;
            message["indicator_left"] = false;
            message["indicator_right_enable"] = false;
            message["indicator_right"] = false;
            message["gear_position_enable"] = false;
            message["gear_position"] = 0;
            message["ebrake_status_enable"] = false;
            message["ebrake_status"] = 0;
            message["remote_drive_ready"] = false;
            message["remote_drive_takeover_request"] = false;
            message["a1_timestamp"] = a1_timestamp;
            
            Json::StreamWriterBuilder writer;
            std::string payload = Json::writeString(writer, message);
            
            // 发送消息
            send_mqtt_message(payload, topic_);
            
            // 记录到CSV队列
            {
                std::lock_guard<std::mutex> lock(csv_mutex_);
                csv_queue_.push({std::to_string(sequence_id_), std::to_string(a1_timestamp)});
            }
            csv_cv_.notify_one();
            
        } catch (const std::exception& e) {
            std::cout << "Error sending message: " << e.what() << std::endl;
        }
    }
    
    void run() {
        try {
            // 初始化MQTT库
            mosquitto_lib_init();
            
            // 创建MQTT客户端
            mosq = mosquitto_new(nullptr, true, nullptr);
            if (!mosq) {
                std::cout << "无法创建MQTT客户端" << std::endl;
                return;
            }
            
            // 设置MQTT回调
            mosquitto_connect_callback_set(mosq, mqtt_connect_callback);
            mosquitto_publish_callback_set(mosq, mqtt_publish_callback);
            mosquitto_disconnect_callback_set(mosq, mqtt_disconnect_callback);
            
            // 连接到MQTT服务器
            int result = mosquitto_connect(mosq, broker_.c_str(), port_, 60);
            if (result != MOSQ_ERR_SUCCESS) {
                std::cout << "MQTT连接失败: " << mosquitto_strerror(result) << std::endl;
                mosquitto_destroy(mosq);
                mosquitto_lib_cleanup();
                return;
            }
            
            // 启动MQTT循环
            mosquitto_loop_start(mosq);
            
            // 启动MQTT发送线程
            std::thread mqtt_thread(mqtt_sender_thread);
            
            std::cout << "MQTT发送程序启动成功" << std::endl;
            std::cout << "发送频率: " << (1.0 / interval_) << "Hz" << std::endl;
            std::cout << "按 Ctrl+C 停止程序" << std::endl;
            
            // 主循环
            while (running_) {
                auto start_time = std::chrono::high_resolution_clock::now();
                send_message();
                auto send_duration = std::chrono::high_resolution_clock::now() - start_time;
                auto sleep_time = std::chrono::duration<double>(interval_) - send_duration;
                if (sleep_time.count() > 0) {
                    std::this_thread::sleep_for(sleep_time);
                }
            }
            
            // 等待MQTT线程结束
            if (mqtt_thread.joinable()) {
                mqtt_thread.join();
            }
            
            // 清理MQTT资源
            mosquitto_loop_stop(mosq, true);
            mosquitto_disconnect(mosq);
            mosquitto_destroy(mosq);
            mosquitto_lib_cleanup();
            
        } catch (const std::exception& e) {
            std::cout << "Error in run loop: " << e.what() << std::endl;
        }
    }
};

int main() {
    // 注册信号处理
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    MQTTSender sender;
    sender.run();
    
    std::cout << "程序已停止" << std::endl;
    return 0;
} 

File: mqtt/mqtt_sender.py
import paho.mqtt.client as mqtt
import json
import time
import signal
import sys
from datetime import datetime
import os
from queue import Queue
from threading import Thread
import csv

class MQTTSender:
    def __init__(self, broker="hellorobotaxi.cn", port=11883, topic="cloud/vehicle/control"):
        self.broker = broker
        self.port = port
        self.topic = topic
        self.client = mqtt.Client()
        self.connected = False
        self.running = True
        self.sequence_id = 0
        self.interval = 0.01  # 10ms
        
        # 创建CSV数据队列
        self.csv_queue = Queue()
        
        # 确保logs目录存在
        if not os.path.exists("logs"):
            os.makedirs("logs")
            
        # 创建带时间戳的日志目录
        timestamp = time.strftime('%Y%m%d_%H%M%S')
        self.log_dir = f"logs/sender_{timestamp}"
        while os.path.exists(self.log_dir):  # 避免目录冲突
            time.sleep(0.1)
            timestamp = time.strftime('%Y%m%d_%H%M%S')
            self.log_dir = f"logs/sender_{timestamp}"
        os.makedirs(self.log_dir)
        
        # 创建并初始化CSV文件
        self.csv_file = os.path.join(self.log_dir, "sender_data.csv")
        with open(self.csv_file, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(['sequence_id', 'a1_timestamp'])
        
        # 启动CSV写入线程
        self.csv_thread = Thread(target=self._csv_writer, daemon=True)
        self.csv_thread.start()

    def on_connect(self, client, userdata, flags, rc):
        self.connected = rc == 0

    def get_current_timestamp(self):
        # 使用系统时间生成毫秒级Unix时间戳
        timestamp_ms = int(time.time() * 1000)
        return timestamp_ms

    def send_message(self):
        if not self.connected:
            return
        
        try:
            # 获取时间戳并发送消息
            self.sequence_id += 1
            a1_timestamp = self.get_current_timestamp()
            
            # 使用与publisher_template.cpp相同的消息格式，但将timestamp替换为a1_timestamp
            message = {
                "sequence_id": self.sequence_id,
                "steering_angle_enable": True,
                "steering_angle": 0.0,
                "target_acceleration_enable": True,
                "target_acceleration": 0.0,
                "indicator_left_enable": False,
                "indicator_left": False,
                "indicator_right_enable": False,
                "indicator_right": False,
                "gear_position_enable": False,
                "gear_position": 0,
                "ebrake_status_enable": False,
                "ebrake_status": 0,
                "remote_drive_ready": False,
                "remote_drive_takeover_request": False,
                "a1_timestamp": a1_timestamp  # 直接使用a1_timestamp替换timestamp
            }
            
            # 发送消息
            payload = json.dumps(message, ensure_ascii=False)
            self.client.publish(self.topic, payload, qos=1)
            
            # 每次发送后立即刷新
            self.client.loop_write()  # 立即写入网络
            
            # 记录sequence_id和a1_timestamp
            self.csv_queue.put([self.sequence_id, a1_timestamp])
            
        except Exception as e:
            print(f"Error sending message: {e}")

    def _csv_writer(self):
        buffer = []
        max_buffer = 100  # 缓冲100条数据再写入
        
        try:
            f = open(self.csv_file, 'a', newline='')
            writer = csv.writer(f)
            
            while True:
                try:
                    data = self.csv_queue.get()
                    if data is None:
                        if buffer:
                            writer.writerows(buffer)
                            f.flush()
                        break
                    
                    buffer.append(data)
                    if len(buffer) >= max_buffer:
                        writer.writerows(buffer)
                        f.flush()
                        buffer = []
                        
                except Exception as e:
                    print(f"Error writing to CSV: {e}")
                    continue
                    
        finally:
            f.close()

    def run(self):
        try:
            self.client.on_connect = self.on_connect
            self.client.connect(self.broker, self.port, 60)
            self.client.loop_start()

            while self.running:
                start_time = time.perf_counter()
                self.send_message()
                send_duration = time.perf_counter() - start_time
                sleep_time = max(0, self.interval - send_duration)
                if sleep_time > 0:
                    time.sleep(sleep_time)
        except Exception as e:
            print(f"Error in run loop: {e}")
            self.cleanup()

    def cleanup(self):
        self.running = False
        self.csv_queue.put(None)
        self.csv_thread.join(timeout=1)
        if self.connected:
            self.client.loop_stop()
            self.client.disconnect()

def signal_handler(signum, frame):
    if mqtt_sender:
        mqtt_sender.running = False

if __name__ == "__main__":
    mqtt_sender = None
    try:
        signal.signal(signal.SIGINT, signal_handler)
        signal.signal(signal.SIGTERM, signal_handler)
        
        mqtt_sender = MQTTSender()
        mqtt_sender.run()
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)
    finally:
        if mqtt_sender:
            mqtt_sender.cleanup()
        sys.exit(0)